test/unit/test_manifest_download.py

```diff
@@ -0,0 +1,75 @@
+import os
+import shutil
+import tempfile
+import unittest
+
+from mock import patch
+
+from hca.dss import DSSClient
+
+
+def _touch_file(*args, **kwargs):
+    path = args[2] if len(args) > 2 else args[0]
+    try:
+        os.makedirs(os.path.split(path)[0])
+    except OSError:
+        # dirs already exist
+        pass
+    with open(path, 'w'):
+        pass
+
+
+class TestManifestDownload(unittest.TestCase):
+    manifest = [
+        ['bundle_uuid', 'bundle_version', 'file_content_type', 'file_name', 'file_sha256', 'file_size', 'file_uuid', 'file_version', 'file_indexed'],
+        ['a_uuid', '1_version', 'somestuff', 'a_file_name', 'ad3fc1e4898e0bce096be5151964a81929dbd2a92bd5ed56a39a8e133053831d', '12', 'af_uuid', 'af_version', 'False'],
+        ['b_uuid', '1_version', 'somestuff', 'b_file_name', '8f35071eaeedd9d6f575a8b0f291daeac4c1dfdfa133b5c561232a00bf18c4b4', '2', 'bf_uuid', 'af_version', 'False'],
+        ['c_uuid', '1_version', 'somestuff', 'c_file_name', '895404db04bdede03e9128a4b48599d0ecde5b2e58ed9ce52ce84c3d54a3429c', '41', 'cf_uuid', 'af_version', 'False'],
+    ]
+
+    def setUp(self):
+        self.prev_wd = os.getcwd()
+        self.tmp_dir = tempfile.mkdtemp()
+        os.chdir(self.tmp_dir)
+        self.dss = DSSClient()
+        with open('manifest.tsv', 'w') as f:
+            f.write('\n'.join(['\t'.join(row) for row in self.manifest]))
+
+    def tearDown(self):
+        os.chdir(self.prev_wd)
+        shutil.rmtree(self.tmp_dir)
+
+    @patch('hca.dss.DSSClient.DIRECTORY_NAME_LENGTHS', [1, 2, 3])
+    def test_file_path(self):
+        self.assertRaises(AssertionError, self.dss._file_path, 'a')
+        parts = self.dss._file_path('abcdefghij').split(os.sep)
+        self.assertEqual(parts, ['a', 'bc', 'def', 'abcdefghij'])
+
+    @patch('hca.dss.DSSClient._download_file', side_effect=_touch_file)
+    def test_manifest_download(self, download_func):
+        files_downloaded = 0
+        self.dss.download_manifest_v2('manifest.tsv', 'aws')
+        self.assertEqual(download_func.call_count, len(self.manifest) - 1)
+        # Since files now exist, running again ensures that we avoid unnecessary downloads
+        self.dss.download_manifest_v2('manifest.tsv', 'aws')
+        self.assertEqual(download_func.call_count, len(self.manifest) - 1)
+        # The check above is good, but we also need to check how many files were actually writen
+        num_files = sum(map(lambda tup: len(tup[2]), os.walk('.')))
```

Comment:
I'd like this to be stronger and assert everything, file names and paths. Not just counts. Also 

```suggestion
        num_files = sum(len(files) for path, dirs, files in os.walk('.'))
```

map() is rarely more readable than a comprehension. And shorter, too. Mine is only longer because I mention the other two tuple elements. Even the version that doesn't name them is more readable, and on par re length.

```suggestion
        num_files = sum(len(files) for _, _, files in os.walk('.'))
```

Reply:
Yes, this is clearly superior.

Reply:
I'd use scandir instead of `os.walk()` if you haven't heard of it yet.  It's a dependency already in `dcp-cli` and it's a part of the os standard library in python3.  It's about 2-20 times faster than walk and seems strictly better when used: https://www.python.org/dev/peps/pep-0471/  The change to the code would be:
`num_files = sum(len(files) for _, _, files in scandir.walk('.'))`

Reply:
Good to know, thanks.

Reply:
It's an order of magnitude increase in speed, which is what the epic is all about.  If you want a blunt experience of it, try it from your home dir vs. `os.walk()`.

Reply:
I believe you, that's why I changed it.